nixos/modules/security/apparmor.nix

```diff
@@ -42,22 +51,55 @@ in
         description = ''
           AppArmor policies.
         '';
-        type = types.attrsOf (types.submodule ({ name, config, ... }: {
-          options = {
-            enable = mkDisableOption "loading of the profile into the kernel";
-            enforce = mkDisableOption "enforcing of the policy or only complain in the logs";
-            profile = lib.mkOption {
-              description = "The policy of the profile.";
-              type = types.lines;
-              apply = pkgs.writeText name;
-            };
-          };
-        }));
-        default = {};
+        type = types.attrsOf (
+          types.submodule (
+            {
+              name,
+              config,
+              options,
+              ...
+            }:
+            {
+              options = {
+                state = lib.mkOption {
+                  description = "The state of the profile as applied to the system by nix";
+                  type = types.enum [
+                    "disable"
+                    "complain"
+                    "enforce"
+                  ];
+                  # should enforce really be the default?
+                  # the docs state that this should only be used once one is REALLY sure nothing's gonna break
+                  default = "enforce";
+                };
+
+                profile = lib.mkOption {
+                  description = "The policy of the profile. Incompatible with path.";
+                  type = types.lines;
+                };
+
+                path = lib.mkOption {
+                  type = types.nullOr types.path;
+                  default = null;
+                  description = "A path of a profile to include. Incompatible with profile.";
+                  apply =
+                    p:
+                    assert (
+                      lib.assertMsg (
+                        (p != null && !options.profile.isDefined) || (p == null && options.profile.isDefined)
+                      ) "`security.apparmor.policies.\"${name}\"` must define exactly one of either path or profile."
```

Comment:
alternatively you could use the assertion mechanism of nixos https://nixos.org/manual/nixos/unstable/#sec-assertions

Reply:
my understanding is assertions do really not like submodules. Am i mistaken here?

Reply:
honestly I don't know

Reply:
> Am i mistaken here?

This is correct, but I think there is a better solution here

Since assertions don't really exist in submodule, we can instead create these assertions from the top-level context by mapping over each policy (we can simplify this assertion with an XOR operator)

```nix
{
  config = {
    assertions = lib.mapAttrsToList (policyName: policyCfg: {
      assertion = lib.xor policyCfg.profile.isDefined (policyCfg.path != null);
      message = "`security.apparmor.policies.\"${policyName}\"` must define exactly one of either path or profile.";
    }) cfg.policies;
  };
}
```

Reply:
Alright, i implemented the assert on the global module level now. You have a point, and i do actually like this way better.